service/worker/scheduler/workflow.go

```diff
@@ -0,0 +1,834 @@
+	s.Info.ActionCount++
+	s.Info.RecentActions = append(s.Info.RecentActions, result)
+	extra := len(s.Info.RecentActions) - 10
+	if extra > 0 {
+		s.Info.RecentActions = s.Info.RecentActions[extra:]
+	}
+	if result.StartWorkflowResult != nil {
+		s.Info.RunningWorkflows = append(s.Info.RunningWorkflows, result.StartWorkflowResult)
+	}
+}
+
+func (s *scheduler) startWorkflow(
+	start *schedspb.BufferedStart,
+	newWorkflow *workflowpb.NewWorkflowExecutionInfo,
+) (*schedpb.ScheduleActionResult, error) {
+	// must match AppendedTimestampForValidation
+	nominalTimeSec := start.NominalTime.UTC().Truncate(time.Second)
+	workflowID := newWorkflow.WorkflowId + "-" + nominalTimeSec.Format(time.RFC3339)
+
+	scheduleToCloseTimeout := 60 * time.Second
+	if !start.Manual {
+		// use catchup window to set timeout
+		deadline := start.ActualTime.Add(s.getCatchupWindow())
+		scheduleToCloseTimeout = deadline.Sub(s.now())
+	}
+	if scheduleToCloseTimeout < 5*time.Second {
+		scheduleToCloseTimeout = 5 * time.Second
+	}
+
+	ctx := workflow.WithActivityOptions(s.ctx, workflow.ActivityOptions{
+		ScheduleToCloseTimeout: scheduleToCloseTimeout,
+		RetryPolicy:            defaultActivityRetryPolicy,
+	})
+
+	// this timestamp is the one used to set the deadline for workflow execution timeout
+	startTime := timestamp.TimePtr(s.now())
+	if !start.Manual {
+		// use the scheduled time for workflow execution timeout even if we started late
+		// TODO: is this being too smart? should we just leave it as the current time?
+		startTime = start.ActualTime
+	}
+
+	req := &schedspb.StartWorkflowRequest{
+		NamespaceId: s.State.NamespaceId,
+		Request: &workflowservice.StartWorkflowExecutionRequest{
+			Namespace:                s.State.Namespace,
+			WorkflowId:               workflowID,
+			WorkflowType:             newWorkflow.WorkflowType,
+			TaskQueue:                newWorkflow.TaskQueue,
+			Input:                    newWorkflow.Input,
+			WorkflowExecutionTimeout: newWorkflow.WorkflowExecutionTimeout,
+			WorkflowRunTimeout:       newWorkflow.WorkflowRunTimeout,
+			WorkflowTaskTimeout:      newWorkflow.WorkflowTaskTimeout,
+			Identity:                 s.identity(),
+			RequestId:                s.newUUIDString(),
+			WorkflowIdReusePolicy:    enumspb.WORKFLOW_ID_REUSE_POLICY_ALLOW_DUPLICATE,
+			RetryPolicy:              newWorkflow.RetryPolicy,
+			Memo:                     newWorkflow.Memo,
+			SearchAttributes:         s.addSearchAttrs(newWorkflow.SearchAttributes, nominalTimeSec),
+			Header:                   newWorkflow.Header,
+		},
+		StartTime:            startTime,
+		LastCompletionResult: s.State.LastCompletionResult,
+		ContinuedFailure:     s.State.ContinuedFailure,
+	}
+	var res schedspb.StartWorkflowResponse
+	// TODO: should this be a local activity instead? or should it be async?
+	err := workflow.ExecuteActivity(ctx, s.a.StartWorkflow, req).Get(s.ctx, &res)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
yes, use local activity, and it should be sync.

Reply:
there's one more nice advantage of local activity here: the inputs won't end up repeated in the history for each start